In [2]:
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})


In [3]:
df


name  math  english  reading
0     Sally    62       85       80
1      Jane    88       79       67
2     Suzie    94       74       95
3     Billy    98       96       88
4       Ada    77       92       98
5      John    79       76       93
6    Thomas    82       64       81
7     Marie    93       63       90
8    Albert    92       62       87
9   Richard    69       80       94
10    Isaac    92       99       93
11     Alan    92       62       72

In [4]:
from pydataset import data

In [15]:
df['passing_english'] = df.reading > 69

In [16]:
df

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67            False
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81             True
7     Marie    93       63       90             True
8    Albert    92       62       87             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
11     Alan    92       62       72             True

In [17]:
df.sort_values(by='passing_english')

name  math  english  reading  passing_english
1      Jane    88       79       67            False
0     Sally    62       85       80             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81             True
7     Marie    93       63       90             True
8    Albert    92       62       87             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
11     Alan    92       62       72             True

Sort the english grades first by passing_english and then by student name.
All the students that are failing english should be first, and within the 
students that are failing english they should be ordered alphabetically. 
The same should be true for the students passing english.
(Hint: you can pass a list to the .sort_values method)

In [20]:
df.sort_values(['passing_english','name'], ascending=[True, True])

name  math  english  reading  passing_english
1      Jane    88       79       67            False
4       Ada    77       92       98             True
11     Alan    92       62       72             True
8    Albert    92       62       87             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True
5      John    79       76       93             True
7     Marie    93       63       90             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
2     Suzie    94       74       95             True
6    Thomas    82       64       81             True

Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.

In [22]:
df.sort_values(['passing_english','english'], ascending=[True,False])

name  math  english  reading  passing_english
1      Jane    88       79       67            False
10    Isaac    92       99       93             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
0     Sally    62       85       80             True
9   Richard    69       80       94             True
5      John    79       76       93             True
2     Suzie    94       74       95             True
6    Thomas    82       64       81             True
7     Marie    93       63       90             True
8    Albert    92       62       87             True
11     Alan    92       62       72             True

Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.

In [25]:
df['overall_grade'] = ((df.math + df.english + df.reading) / 3).astype(int)

In [26]:
df

name  math  english  reading  passing_english  overall_grade
0     Sally    62       85       80             True             75
1      Jane    88       79       67            False             78
2     Suzie    94       74       95             True             87
3     Billy    98       96       88             True             94
4       Ada    77       92       98             True             89
5      John    79       76       93             True             82
6    Thomas    82       64       81             True             75
7     Marie    93       63       90             True             82
8    Albert    92       62       87             True             80
9   Richard    69       80       94             True             81
10    Isaac    92       99       93             True             94
11     Alan    92       62       72             True             75

In [27]:
mpg = data('mpg')

In [28]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




In [29]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


How many rows and columns are there?

What are the data types of each column?


-234 rows and 11 columns
-object,object,float,int,int,object,object,int,int,object,object

In [30]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

Rename the cty column to city.

In [38]:
mpg.rename(columns={'cty': 'city'})


manufacturer   model  displ  year  cyl       trans drv  city  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16   26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

Rename the hwy column to highway.


In [39]:
mpg.rename(columns={'hwy': 'highway'})

manufacturer   model  displ  year  cyl       trans drv  cty  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16       26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

Do any cars have better city mileage than highway mileage?


In [42]:
mpg[mpg.cty > mpg.hwy]

Empty DataFrame
Columns: [manufacturer, model, displ, year, cyl, trans, drv, cty, hwy, fl, class]
Index: []

Create a column named mileage_difference this column 
should contain the difference between highway and city mileage for each 
car.


In [46]:
mpg['mileage_difference'] = mpg.hwy - mpg.cty

In [47]:
mpg


manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_differnce  mileage_difference  
1    compact                 11                  11  
2    compact                  8                   8  
3    compact                 11                  11  
4    compact                  9                   9  
5    compact                 10                  10  
..       ...                ...                 ...  
230  midsize                  9                   9  
231  midsize                  8                   8  
232  midsize                 10                  10  
233  midsize                  8                   8  
234  midsize                  9                   9  

[234 rows x 13 columns]

Which car (or cars) has the highest mileage difference?


In [56]:
highest = mpg.mileage_difference.max()

In [58]:
mpg[mpg.mileage_difference == highest]

manufacturer       model  displ  year  cyl     trans drv  cty  hwy fl  \
107        honda       civic    1.8  2008    4  auto(l5)   f   24   36  c   
223   volkswagen  new beetle    1.9  1999    4  auto(l4)   f   29   41  d   

          class  mileage_differnce  mileage_difference  
107  subcompact                 12                  12  
223  subcompact                 12                  12

Which compact class car has the lowest highway mileage? The best?


In [60]:
mpg[(mpg.class == "compact") and mpg.hwy.min()]

SyntaxError: invalid syntax (<ipython-input-60-2cd719f32017>, line 1)

Create a column named average_mileage that is the mean of the city and highway mileage.

In [80]:
mpg['average_mileage'] = (mpg.cty + mpg.hwy) / 2

Which dodge car has the best average mileage? The worst?


In [81]:
mpg


manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_differnce  mileage_difference  average_mileage  
1    compact                 11                  11             23.5  
2    compact                  8                   8             25.0  
3    compact                 11                  11             25.5  
4    compact                  9                   9             25.5  
5    compact                 10                  10             21.0  
..       ...                ...                 ...              ...  
230  midsize                  9                   9             23.5  
231  midsize                  8                   8             25.0  
232  midsize                 10                  10             21.0  
233  midsize                  8                   8             22.0  
234  midsize                  9                   9             21.5  

[234 rows x 14 columns]

dodge = mp

In [85]:
dodge = mpg[mpg['manufacturer'] == 'dodge'] and mpg['average_mileage']

ValueError: The truth value of a DataFrame is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().